pkg/templates/nstemplatetiers/nstemplatetier_generator.go

```diff
@@ -23,74 +24,110 @@ import (
-	tierTmplsByTier, err := newTierTemplates(s, namespace, assets)
+	// initialize tier generator, loads templates from assets
+	generator, err := newTierGenerator(s, client, namespace, assets)
 	if err != nil {
 		return errors.Wrap(err, "unable to create TierTemplates")
 	}
-	for _, tierTmpls := range tierTmplsByTier {
-		for _, tierTmpl := range tierTmpls {
-			// using the "standard" client since we don't need to support updates on such resources, they should be immutable
-			if err := client.Create(context.TODO(), tierTmpl); err != nil && !apierrors.IsAlreadyExists(err) {
-				return errors.Wrapf(err, "unable to create the '%s' TierTemplate in namespace '%s'", tierTmpl.Name, tierTmpl.Namespace)
-			}
-			log.Info("TierTemplate resource created", "namespace", tierTmpl.Namespace, "name", tierTmpl.Name)
-		}
+
+	// create the TierTemplate resources
+	err = generator.createTierTemplates()
+	if err != nil {
+		return errors.Wrap(err, "unable to create TierTemplates")
 	}
-	// create the NSTemplateTiers
-	nstmplTiersByTier := newNSTemplateTiers(namespace, tierTmplsByTier)
-	cl := commonclient.NewApplyClient(client, s)
 
-	for _, nstmplTier := range nstmplTiersByTier {
-		createdOrUpdated, err := cl.CreateOrUpdateObject(nstmplTier, true, nil)
-		if err != nil {
-			return errors.Wrapf(err, "unable to create or update the '%s' NSTemplateTiers in namespace '%s'", nstmplTier.Name, nstmplTier.Namespace)
-		}
-		if createdOrUpdated {
-			log.Info("NSTemplateTier resource created/updated", "namespace", nstmplTier.Namespace, "name", nstmplTier.Name)
-		} else {
-			log.Info("NSTemplateTier resource was already up-to-date", "namespace", nstmplTier.Namespace, "name", nstmplTier.Name, "ResourceVersion", nstmplTier.ResourceVersion)
-		}
+	// create the NSTemplateTier resources
+	err = generator.createNSTemplateTiers()
+	if err != nil {
+		return errors.Wrap(err, "unable to create NSTemplateTiers")
 	}
+
 	return nil
 }
 
-// templates: namespaces and other cluster-scoped resources belonging to a given tier ("advanced", "basic", "team", etc.)
+type tierGenerator struct {
+	client          client.Client
+	namespace       string
+	scheme          *runtime.Scheme
+	templatesByTier map[string]*tierData
+}
+
+type tierData struct {
+	rawTemplates   *templates
+	tierTemplates  []*toolchainv1alpha1.TierTemplate
+	nstmplTierObjs []commonclient.ToolchainObject
+}
+
+// templates: namespaces and other cluster-scoped resources belonging to a given tier ("advanced", "basic", "team", etc.) and the NSTemplateTier that combines them
 type templates struct {
 	namespaceTemplates map[string]template // namespace templates (including roles, etc.) indexed by type ("dev", "code", "stage")
 	clusterTemplate    *template           // other cluster-scoped resources, in a single template file
+	nsTemplateTier     *template           // NSTemplateTier resource with tier-scoped configuration and references to namespace and cluster templates in its spec, in a single template file
```

Comment:
`clusterTemplate` is a pointer to `template` because it's optional, but I believe that the new `tier.yaml` template must exist for each tier, so I guess you don't need a pointer to `template` here?

Reply:
That's a good point. I can try changing that.

Reply:
Done in https://github.com/codeready-toolchain/host-operator/pull/269/commits/b571580f5c26cb418c69a9ee10f09565bdd033ca